sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/ViewRegistry.java

```diff
@@ -71,17 +70,79 @@ private static boolean matchesSelector(
       InstrumentSelector selector,
       InstrumentDescriptor descriptor,
       InstrumentationScopeInfo meterScope) {
-    return (selector.getInstrumentType() == null
-            || selector.getInstrumentType() == descriptor.getType())
-        && selector.getInstrumentNameFilter().test(descriptor.getName())
-        && matchesMeter(selector, meterScope);
+    if (selector.getInstrumentType() != null
+        && selector.getInstrumentType() != descriptor.getType()) {
+      return false;
+    }
+    if (selector.getInstrumentName() != null
+        && !matchesName(selector.getInstrumentName(), descriptor.getName())) {
+      return false;
+    }
+    return matchesMeter(selector, meterScope);
+  }
+
+  /**
+   * Determine if the {@code instrumentName} matches the {@code namePattern} from {@link
+   * InstrumentSelector#getInstrumentName()}.
+   *
+   * <p>{@code namePattern} may contain the wildcard characters {@code *} and {@code ?} with the
+   * following matching criteria:
+   *
+   * <ul>
+   *   <li>{@code *} matches 0 or more instances of any character
+   *   <li>{@code ?} matches exactly one instance of any character
+   * </ul>
+   *
+   * <p>Based off implementation from: https://www.rgagnon.com/javadetails/java-0515.html
+   */
+  // Visible for testing
+  static boolean matchesName(String namePattern, String instrumentName) {
+    if (!namePattern.contains("*") && !namePattern.contains("?")) {
+      return namePattern.equals(instrumentName);
+    }
+
+    StringBuilder builder = new StringBuilder();
+    for (int i = 0; i < namePattern.length(); i++) {
+      char c = namePattern.charAt(i);
+      switch (c) {
+        case '*':
+          builder.append(".*");
+          break;
+        case '?':
+          builder.append('.');
+          break;
+        case '(':
+        case ')':
+        case '[':
+        case ']':
+        case '$':
+        case '^':
+        case '.':
+        case '{':
+        case '}':
+        case '|':
+          builder.append("\\").append(c);
+          break;
+        default:
+          builder.append(c);
+          break;
+      }
+    }
+    String regex = builder.append('$').toString();
+    return Pattern.matches(regex, instrumentName);
```

Comment:
Please take a close look at this translation logic from wildcard pattern to regex.

Reply:
it looks ok to me. have you considered building this Pattern in the InstrumentSelector itself, rather than every time its evaluated?

Reply:
Good point - currently, for each instrument, its computed once for each view registered. Definitely should consider caching minimally, even if its not on the hot path.

I considered adding it to InstrumentSelector, but don't love it. We want to know the raw instrument name selection when a view is configured so that we can disallow registering views that change the instrument name and select multiple instrument on the same meter, as [required by the spec](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/sdk.md#L197-L200). Its more difficult to tell if a pattern represents an exact match or has wildcards. This sort of implies having two methods on InstrumentSelector, one for exact name match, and one for a pattern match. I figure its better to limit the API surface area and just have one.

Technically we don't need InstrumentSelector to have any public methods. Instead, it could have a package private method that returns an internal class with getters to all the values. If we did this, then we could add whatever getters we wanted without worrying about API surface area. Same thing could be applied to `View`.

Reply:
> Technically we don't need InstrumentSelector to have any public methods. Instead, it could have a package private method that returns an internal class with getters to all the values.

Although this is inconsistent with our our configuration data classes like `SpanLimits`.

Reply:
yeah, that would be a little weird. It probably doesn't matter much, since this *shouldn't* be on the hot path (although users are gonna user, of course).